### PR TITLE
[RLlib] Add swa option for base torch optimizer

### PR DESCRIPTION
**Pull request by smorad**
## Why are these changes needed?
Stochastic weight averaging (SWA) is known to significantly improve generalization capabilities in supervised learning at very little cost. [Recent works](https://www.gatsby.ucl.ac.uk/~balaji/udl-camera-ready/UDL-24.pdf) suggest it improves stability in RL as well.

This patch introduces `swa` for `torch` models. Using the tuned PPO cartpole example, changes seem significant:
```yaml
# without_swa.yaml
cartpole-ppo:
    env: CartPole-v0
    run: PPO
    stop:
        timesteps_total: 200000
    config:
        # Works for both torch and tf.
        framework: torch
        gamma: 0.99
        lr: 0.0003
        num_workers: 1
        observation_filter: MeanStdFilter
        num_sgd_iter: 6
        vf_loss_coeff: 0.01
        swa: False
        model:
            fcnet_hiddens: [32]
            fcnet_activation: linear
            vf_share_layers: true
```
```yaml
# with_swa.yaml
cartpole-ppo:
    env: CartPole-v0
    run: PPO
    stop:
        timesteps_total: 200000
    config:
        # Works for both torch and tf.
        framework: torch
        gamma: 0.99
        lr: 0.0003
        num_workers: 1
        observation_filter: MeanStdFilter
        num_sgd_iter: 6
        vf_loss_coeff: 0.01
        swa: True
        swa_config:
          start: 0
          # Every 3 batches
          freq: 18
        model:
            fcnet_hiddens: [32]
            fcnet_activation: linear
            vf_share_layers: true
```

```bash
rllib train -f without_swa.yaml
rllib train -f with_swa.yaml
```

The pink line is without SWA and the red line is with SWA. As you can see, SWA seems to greatly improve policy stability. I'm excited to see the results for more difficult tasks.
Screen Shot 2022-02-28 at 12 52 03 PM

Note that this only works for algorithms which use the base optimizer. Policies that instantiate their own optimizers (utilizing `policy_template.py:build_policy` can't use this without plumbing this change through each policy individually. This can be done, but it is a lot of boilerplate so I'll wait to see how everyone feels before starting.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [x] This PR is not tested :(

### CODE REVIEW

File: rllib/policy/torch_policy.py [hunk 1]
```diff
@@ -842,6 +842,16 @@ def optimizer(
             optimizers = [
                 torch.optim.Adam(self.model.parameters(), lr=self.config["lr"])
             ]
+            if self.config["swa"]:
```
Comment (sven1977) [inline rllib/policy/torch_policy.py:845]: This is awesome! Thanks Steven for the research and this PR here.
* Could we add an informative error here iff torchcontrib is not installed?

Comment (smorad) [reply rllib/policy/torch_policy.py:845]: Sure, patch incoming

**Comment by smorad** [comment]
It's not clear to me this is the best spot to place this change. It seems like depending on the algorithm, the optimizers are implemented in different places.

**Comment by duburcqa** [comment]
I get your point and I do agree it is a strong advantage. Yet, I also think the drawbacks I mentioned cannot be ignored. Would it be possible de implement a small wrapper on top of `torch.optim.swa_utils` to embed everything in a wrapper that can be applied to any optimizer ? It is a bit more work but otherwise I don't see in future for this other than prototyping.

**Comment by duburcqa** [comment]
I had a quick look to the source, and I see no issue to implement a wrapper for optimizers. The `AveragedModel` is NOT the model used for inference. So it can be kept internally in the wrapper without issue. No need to wrap the model from the user point of view.

**Comment by duburcqa** [comment]
Ok last update. I missed the point that the `torchcontrib` already support any optimizer apparently, even though it was only reliably tested on `SGD`. So I guess if it works it is fine.

**Comment by smorad** [comment]
No worries. Using `AveragedModel` does indeed fail:
```python
AttributeError: 'AveragedModel' object has no attribute 'get_initial_state'
```
Because all of the `ModelV2` methods such as `value_function`, `get_initial_state`, etc are no longer accessible once we wrap the `ModelV2` with the `AveragedModel`.

**Comment by smorad** [comment]
That is only in manual mode, when `swa_lr` is specified. We use automatic mode:

https://github.com/pytorch/contrib/blob/c545fedf4f73c8e95f91fd81f2d5bf7fa9c62a61/torchcontrib/optim/swa.py#L45-L63

**Comment by duburcqa** [comment]
Hum... no. The doc to are referring to is stated it is for both.

**Comment by smorad** [comment]
I'm not sure.

But here are the results using IMPALA with and without SWA:

Screen Shot 2022-02-28 at 5 05 02 PM

As you can see, there is a significant difference. It must be updating the weights somehow, right?

**Comment by duburcqa** [comment]
I agree it looks to do something but writing the code in fact it is not. So I don't know why the results are different. May be it is just not repeatable.

**Comment by sven1977** [comment]
Hey @smorad , this looks indeed very clear to me for IMPALA! So just to confirm the orange, light-blue, and grey curves are w/ SWA?

**Comment by duburcqa** [comment]
Yet, I'm 100% sure it is not doing anything. It is absolute impossible. Even in the theoretical pseudo-code algorithm it does nothing to the main model. It just generate another set of parameters that if replace the original one by those ones you can expect to get better results. So it makes absolutely no sense.

**Comment by smorad** [comment]
Yeah, the ones on the left that all terminate much sooner are with SWA. But @duburcqa is right, from what I see the buffers are not being swapped with model weights so I'm not sure why this is occurring. Honestly, with just 1-2 runs I would say it's due to variance, but over ~10 runs there is obviously a trend.

**Comment by smorad** [comment]
I'm closing this in favor of #22765, which uses @duburcqa's suggestions of applying SWA to the model instead of the optimizer.

**Pull request closed without merging**